### PR TITLE
second attempt to add CAVLC B-Frame decoding support.

### PR DESCRIPTION
**Pull request by xiaotianshi2**

### CODE REVIEW

File: codec/console/common/src/read_config.cpp [hunk 1]
```diff
@@ -37,8 +37,9 @@
  *      08/18/2008 Created
  *
  *****************************************************************************/
-
+#if !defined(_CRT_SECURE_NO_WARNINGS)
```
Comment (mstorsjo) [inline codec/console/common/src/read_config.cpp]: This change (and the other similar one) is entirely unrelated to the rest of this pull request. Please refrain from doing unrelated changes in one pull request; separate such changes out to a separate on if they are needed.

File: codec/decoder/core/src/manage_dec_ref.cpp [hunk 2]
```diff
@@ -374,7 +374,7 @@ int32_t WelsReorderRefList (PWelsDecoderContext pCtx) {
           }
           iPredFrameNum &= iMaxPicNum - 1;
 
-          for (i = iMaxRefIdx - 1; i >= 0; i--) {
+          for (i = iMaxRefIdx; i >= 0; i--) {
```
Comment (mstorsjo) [inline codec/decoder/core/src/manage_dec_ref.cpp]: This change looks rather suspicious, although I'm not off-hand familiar enough to comment on it.

File: codec/decoder/core/src/decode_slice.cpp [hunk 3]
```diff
@@ -51,6 +51,13 @@
 #include "cpu_core.h"
 
 namespace WelsDec {
+//#define _MV_DUMP_ 1
+#ifdef _MV_DUMP_
+#if defined(_DEBUG)
+FILE* pFile = fopen ("Y:/p4/CodeProjects/Upwork/screencast/xiaotiansf/openh264-1/MV_CAVLC.txt", "w");
```
Comment (mstorsjo) [inline codec/decoder/core/src/decode_slice.cpp]: I don't think this kind of leftover debug code should be included.

File: codec/decoder/core/src/parse_mb_syn_cavlc.cpp [hunk 4]
```diff
@@ -46,6 +46,13 @@
 
 namespace WelsDec {
 #define MAX_LEVEL_PREFIX 15
+//#define _MV_DUMP_ 1
```
Comment (mstorsjo) [inline codec/decoder/core/src/parse_mb_syn_cavlc.cpp]: Please leave out the debug code.

File: test/api/encode_options_test.cpp [hunk 5]
```diff
@@ -1033,6 +1033,13 @@ TEST_F (EncodeDecodeTestAPI, SimulcastSVC) {
       EXPECT_TRUE (iResult == cmResultSuccess) << "iResult=" << iResult << "LayerIdx=" << iIdx;
       iResult = decoder[iIdx]->DecodeFrame2 (NULL, 0, pData, &dstBufInfo_);
       EXPECT_TRUE (iResult == cmResultSuccess) << "iResult=" << iResult << "LayerIdx=" << iIdx;
+      if (dstBufInfo_.iBufferStatus == 0) {
```
Comment (mstorsjo) [inline test/api/encode_options_test.cpp]: Why these changes, in the encoding options test, when you are implementing a new mode in the decoder? This feels unrelated.

File: test/decoder/DecUT_ParseSyntax.cpp [hunk 6]
```diff
@@ -412,7 +412,7 @@ void DecoderParseSyntaxTest::TestSpecificBsError() {
   int32_t iRet = ERR_NONE;
   Init();
   ASSERT_EQ (iRet, ERR_NONE);
-  ASSERT_TRUE (DecodeBs ("res/Cisco_Men_whisper_640x320_CAVLC_Bframe_9.264", ErrorDec));
+  ASSERT_TRUE (DecodeBs ("res/Cisco_Men_whisper_640x320_CAVLC_Bframe_9.264", CorrectDec));
```
Comment (mstorsjo) [inline test/decoder/DecUT_ParseSyntax.cpp]: Is this needed and relevant for the new feature?

File: codec/console/common/src/read_config.cpp (hunk 1 above)
Comment (xiaotianshi2) [reply codec/console/common/src/read_config.cpp]: I see.
Thanks for pointing out..

File: codec/decoder/core/src/manage_dec_ref.cpp (hunk 2 above)
Comment (xiaotianshi2) [reply codec/decoder/core/src/manage_dec_ref.cpp]: I did have the same doubt until I reviewed the ffmpeg source code and JM's reference decoder. 
This iMaxRefIdx - 1 had given a lot of trouble on several test streams and finally I realized that the start index must start from iMaxRefIdx. I'll certainly find more test cases to prove it.

File: codec/decoder/core/src/decode_slice.cpp (hunk 3 above)
Comment (xiaotianshi2) [reply codec/decoder/core/src/decode_slice.cpp]: I'll remove it as soon as I done.

File: test/api/encode_options_test.cpp (hunk 5 above)
Comment (xiaotianshi2) [reply test/api/encode_options_test.cpp]: This is because there is no way to determine if there is B-frame to be present in the stream. So the ready decoded images have to be queued for reordering before they can be for output..

File: test/decoder/DecUT_ParseSyntax.cpp (hunk 6 above)
Comment (xiaotianshi2) [reply test/decoder/DecUT_ParseSyntax.cpp]: I guess so.

**Comment by xiaotianshi2** [comment]
@thaytan,
Could you provide me the sample asset CAWP5_TOSHIBA_E.264 file from the AVCv1 conformance set?
Thanks.

**Comment by xiaotiansf** [comment]
@thaytan,
I reverted my changes in manage_def_ref.cpp.
Could you try file CAWP5_TOSHIBA_E.264 again with the following change in manage_def_ref.cpp?
line 349 "int32_t iMaxRefIdx = pCtx->iPicQueueNumber;"
Thanks.

**Comment by thaytan** [comment]
> @thaytan,
> Could you provide me the sample asset CAWP5_TOSHIBA_E.264 file from the AVCv1 conformance set?
> Thanks.

All the AVC conformance bitstreams are available at https://www.itu.int/net/itu-t/sigdb/spevideo/VideoForm-s.aspx?val=102002641

**Comment by raypat56** [comment]
Fuck the operation

RPG

**Comment by thaytan** [comment]
Is this PR obsolete now that you filed PR 3060?

**Comment by thaytan** [comment]
> @thaytan,
> I reverted my changes in manage_def_ref.cpp.
> Could you try file CAWP5_TOSHIBA_E.264 again with the following change in manage_def_ref.cpp?
> line 349 "int32_t iMaxRefIdx = pCtx->iPicQueueNumber;"
> Thanks.

It still segfaults with that change.

**Comment by xiaotianshi2** [comment]
@thaytan thanks for verification.

**Comment by sijchen** [comment]
close this for failed CI history.

**Comment by thaytan** [comment]
> @thaytan I am continuing to working on the multi-reference reordering issue. I have tried CAWP5_TOSHIBA_E.264 on Linux with my new changes but I never gets the segfaults. I also compared the dumped reordered ref picture lists to JM's reference decoder's one. They now match perfectly. Could you try my new branch https://github.com/xiaotianshicomcast/openh264/tree/MultiRefTempDirectFix to see if you still see the segfaults? If yes, could you let me know your system configurations in order to reproduce the segfaults so that I can also reproduce the segfaults? Thanks.

I don't get segfaults with that branch.
